packages/flutter_tools/lib/src/android/native_assets.dart

```diff
@@ -0,0 +1,243 @@
+  }
+
+  final List<Target> targets = androidArchs.map(_getNativeTarget).toList();
+  final native_assets_cli.BuildMode buildModeCli =
+      nativeAssetsBuildMode(buildMode);
+
+  globals.logger
+      .printTrace('Building native assets for $targets $buildModeCli.');
+  final List<Asset> nativeAssets = <Asset>[];
+  final Set<Uri> dependencies = <Uri>{};
+  for (final Target target in targets) {
+    final BuildResult result = await buildRunner.build(
+      linkModePreference: LinkModePreference.dynamic,
+      target: target,
+      buildMode: buildModeCli,
+      workingDirectory: projectUri,
+      includeParentEnvironment: true,
+      cCompilerConfig: await buildRunner.ndkCCompilerConfig,
+      targetAndroidNdkApi: targetAndroidNdkApi,
+    );
+    ensureNativeAssetsBuildSucceed(result);
+    nativeAssets.addAll(result.assets);
+    dependencies.addAll(result.dependencies);
+  }
+  ensureNoLinkModeStatic(nativeAssets);
+  globals.logger.printTrace('Building native assets for $targets done.');
+  if (isAndroidLibrary && nativeAssets.isNotEmpty) {
+    throwToolExit('Native assets are not yet supported in Android add2app.');
+  }
+  final Map<Asset, Asset> assetTargetLocations =
+      _assetTargetLocations(nativeAssets);
+  await _copyNativeAssetsAndroid(buildUri_, assetTargetLocations, fileSystem);
+  final Uri nativeAssetsUri = await writeNativeAssetsYaml(
+      assetTargetLocations.values,
+      yamlParentDirectory ?? buildUri_,
+      fileSystem);
+  return (nativeAssetsUri, dependencies.toList());
+}
+
+Future<void> _copyNativeAssetsAndroid(
+  Uri buildUri,
+  Map<Asset, Asset> assetTargetLocations,
+  FileSystem fileSystem,
+) async {
+  if (assetTargetLocations.isNotEmpty) {
+    globals.logger
+        .printTrace('Copying native assets to ${buildUri.toFilePath()}.');
+    final List<String> jniArchDirs = <String>[
+      for (final AndroidArch androidArch in AndroidArch.values)
+        androidArch.archName,
+    ];
+    for (final String jniArchDir in jniArchDirs) {
+      final Uri archUri = buildUri.resolve('jniLibs/lib/$jniArchDir/');
+      await fileSystem.directory(archUri).create(recursive: true);
+    }
+    for (final MapEntry<Asset, Asset> assetMapping
+        in assetTargetLocations.entries) {
+      final Uri source = (assetMapping.key.path as AssetAbsolutePath).uri;
+      final Uri target = (assetMapping.value.path as AssetAbsolutePath).uri;
+      final AndroidArch androidArch =
+          _getAndroidArch(assetMapping.value.target);
+      final String jniArchDir = androidArch.archName;
+      final Uri archUri = buildUri.resolve('jniLibs/lib/$jniArchDir/');
+      final Uri targetUri = archUri.resolveUri(target);
+      final String targetFullPath = targetUri.toFilePath();
+      await fileSystem.file(source).copy(targetFullPath);
+    }
+    globals.logger.printTrace('Copying native assets done.');
+  }
+}
+
+/// Get the [Target] for [androidArch].
+Target _getNativeTarget(AndroidArch androidArch) {
```

Comment:
Should this method (and/or _getAndroidArch) be static?

Reply:
These are toplevel methods, not nested in a class.